File: LC/[1340] Jump Game V - DP.cpp
class Solution {
public:
    int maxJumps(vector<int>& arr, int d) {
        int N = arr.size();

        //declare and initialize the DP
        vector<int>dp(N, 1);

        //sort the array        
        vector<pair< int, int>>p;
        for(int ii=0; ii<N; ++ii){
            p.push_back({arr[ii], ii});
        }

        sort(p.begin(), p.end());//sort by height ascending
        reverse(p.begin(), p.end());//sorted by height decending

        for(int ii=0; ii<N; ++ii){
            int idx = p[ii].second;
            for(int jj=idx+1; jj<=min(N-1, idx+d); ++jj){
                if(arr[jj] >= arr[idx]) break; //cannot have anything higher in between
                dp[jj] = max(dp[jj], dp[idx]+1);
            }

            for(int jj=idx-1; jj>=max(0, idx-d); --jj){
                if(arr[jj] >= arr[idx]) break; //cannot have anything higher in between
                dp[jj] = max(dp[jj], dp[idx]+1);
            }
        }

        int ret = 0;
        for(int ii=0; ii<N; ++ii){
            ret = max(ret, dp[ii]);
        }
        return ret;
    }
};

File: LC/[1340] Jump Game V - DP.java
import java.util.*;

class Solution {
    public int maxJumps(int[] arr, int d) {
        int n = arr.length;
        int[] dp = new int[n];
        Arrays.fill(dp, 1);

        // Sort indices by height ascending (compute from lower to higher)
        Integer[] idx = new Integer[n];
        for (int i = 0; i < n; i++) idx[i] = i;
        Arrays.sort(idx, (a, b) -> arr[a] != arr[b] ? Integer.compare(arr[a], arr[b]) : Integer.compare(a, b));

        for (int k = 0; k < n; k++) {
            int i = idx[k];

            // Look right
            for (int j = i + 1; j <= Math.min(n - 1, i + d); j++) {
                if (arr[j] >= arr[i]) break;                 // blocked
                dp[i] = Math.max(dp[i], 1 + dp[j]);          // pull from lower
            }
            // Look left
            for (int j = i - 1; j >= Math.max(0, i - d); j--) {
                if (arr[j] >= arr[i]) break;                 // blocked
                dp[i] = Math.max(dp[i], 1 + dp[j]);          // pull from lower
            }
        }

        int ans = 0;
        for (int v : dp) ans = Math.max(ans, v);
        return ans;
    }
}